File: Polars.CSharp/Operations/Sort.cs
using System.Data;
using Polars.NET.Core;
using Pl = Polars.CSharp.Polars;

namespace Polars.CSharp;
public partial class LazyFrame : IDisposable, IPolarsLazyFrame
{
    /// <summary>
    /// Sort the LazyFrame by a single column name, selector, or column expression.
    /// </summary>
    public LazyFrame Sort(
        IntoSelector by, 
        bool descending = false, 
        bool nullsLast = false, 
        bool maintainOrder = false)
    {
        using var safeSelector = by.Consume();
        using var expr = safeSelector.ToExpr();
        
        var handles = new[] { PolarsWrapper.CloneExpr(expr.Handle) };
        var h = PolarsWrapper.LazyFrameSort(CloneHandle(), handles, [descending], [nullsLast], maintainOrder);
        
        return new LazyFrame(h);
    }
    /// <summary>
    /// Lazily sort the DataFrame by multiple columns.
    /// <para>
    /// This operation is added to the logical plan. 
    /// Use <see cref="LazyFrame.TopK(int, IEnumerable{Expr}, bool[])"/> if you only need the top/bottom N rows, as it is more efficient.
    /// </para>
    /// </summary>
    /// <param name="by">Names of the columns to sort by.</param>
    /// <param name="descending">Sort order for each column.</param>
    /// <param name="nullsLast">Whether nulls go last for each column.</param>
    /// <param name="maintainOrder">Whether to maintain the relative order of rows with equal keys.</param>
    /// <example>
    /// <code>
    /// df.Lazy()
    ///   .Sort(
    ///       by: new[] { "group", "val" }, 
    ///       descending: new[] { false, true }, 
    ///       nullsLast: new[] { false, false }
    ///   )
    ///   .Collect();
    /// /* Output:
    /// shape: (5, 2)
    /// ┌───────┬─────┐
    /// │ group ┆ val │
    /// │ ---   ┆ --- │
    /// │ str   ┆ i32 │
    /// ╞═══════╪═════╡
    /// │ A     ┆ 10  │
    /// │ A     ┆ 8   │
    /// │ ...   ┆ ... │
    /// └───────┴─────┘
    /// */
    /// </code>
    /// </example>
    public LazyFrame Sort(
        IEnumerable<IntoSelector> by, 
        bool[]? descending = null, 
        bool[]? nullsLast = null, 
        bool maintainOrder = false)
    {
        var selectors = by as IntoSelector[] ?? [.. by];
        if (selectors.Length == 0) return this;

        var handles = new ExprHandle[selectors.Length];
        for (int i = 0; i < selectors.Length; i++)
        {
            using var sel = selectors[i].Consume();
            using var expr = sel.ToExpr();
            handles[i] = PolarsWrapper.CloneExpr(expr.Handle);
        }

        var desc = descending ?? [false];
        var nulls = nullsLast ?? [false];

        var h = PolarsWrapper.LazyFrameSort(CloneHandle(), handles, desc, nulls, maintainOrder);
        return new LazyFrame(h);
    }

    /// <summary>
    /// Sort the LazyFrame using specific sorting configurations for each column/selector.
    /// </summary>
    public LazyFrame Sort(params (IntoSelector By, bool Descending)[] sortConfigs)
    {
        if (sortConfigs.Length == 0) return this;

        var handles = new ExprHandle[sortConfigs.Length];
        var descArray = new bool[sortConfigs.Length];
        var nullsArray = new bool[sortConfigs.Length]; 

        for (int i = 0; i < sortConfigs.Length; i++)
        {
            using var sel = sortConfigs[i].By.Consume();
            using var expr = sel.ToExpr();
            
            handles[i] = PolarsWrapper.CloneExpr(expr.Handle);
            descArray[i] = sortConfigs[i].Descending;
            nullsArray[i] = false; 
        }

        var h = PolarsWrapper.LazyFrameSort(CloneHandle(), handles, descArray, nullsArray, maintainOrder: false);
        return new LazyFrame(h);
    }
}

public partial class DataFrame : IDisposable,IEnumerable<Series>,IPolarsDataFrame
{
    /// <summary>
    /// Sort the LazyFrame by a single column name, selector, or column expression.
    /// </summary>
    public DataFrame Sort(
        IntoSelector by, 
        bool descending = false, 
        bool nullsLast = false, 
        bool maintainOrder = false)
    {
        return Lazy().Sort(
            by, 
            descending, 
            nullsLast, 
            maintainOrder
        ).Collect();
    }
    /// <summary>
    /// Sort the DataFrame by multiple columns (all ascending or all descending).
    /// </summary>
    public DataFrame Sort(
        IEnumerable<IntoSelector> by, 
        bool[]? descending = null, 
        bool[]? nullsLast = null, 
        bool maintainOrder = false)
    {
        return Lazy().Sort(
            by, 
            descending, 
            nullsLast, 
            maintainOrder
        ).Collect();
    }
    /// <summary>
    /// Sort the LazyFrame using specific sorting configurations for each column/selector.
    /// </summary>
    public DataFrame Sort(params (IntoSelector By, bool Descending)[] sortConfigs)
        => Lazy().Sort(sortConfigs).Collect();

}

File: Polars.FSharp/Operations/Sort.fs
namespace Polars.FSharp
open Polars.NET.Core

[<AutoOpen>]
module SortOps = 
    type Series with
        /// <summary>
        /// Sort this Series. Returns a new Series.
        /// </summary>
        /// <param name="descending">Sort in descending order (default: false).</param>
        /// <param name="nullsLast">Place null values last (default: false).</param>
        /// <param name="maintainOrder">Maintain the order of equal elements (Stable sort) (default: false).</param>
        /// <param name="multithreaded">Use multiple threads (default: true).</param>
        member this.Sort(
            ?descending: bool,
            ?nullsLast: bool,
            ?maintainOrder: bool,
            ?multithreaded: bool
        ) =
            let desc = defaultArg descending false
            let nLast = defaultArg nullsLast false
            let stable = defaultArg maintainOrder false
            let multi = defaultArg multithreaded true

            new Series(PolarsWrapper.SeriesSort(this.Handle, desc, nLast, multi, stable))
    type LazyFrame with
        member this.Sort(
            columns: seq<IColumnExpr>, 
            descending: seq<bool>,
            nullsLast: seq<bool>,
            ?maintainOrder: bool
        ): LazyFrame =
            let exprHandles = 
                columns 
                |> Seq.collect (fun x -> x.ToExprs()) 
                |> Seq.map (fun e -> e.CloneHandle()) 
                |> Seq.toArray
            
            let descArr = descending |> Seq.toArray
            let nullsArr = nullsLast |> Seq.toArray
            let stable = defaultArg maintainOrder false

            let lfHandle = this.CloneHandle()
            let h = PolarsWrapper.LazyFrameSort(lfHandle, exprHandles, descArr, nullsArr, stable)
            new LazyFrame(h)
        member this.Sort(
            columns: seq<Expr>, 
            descending: seq<bool>,
            nullsLast: seq<bool>,
            ?maintainOrder: bool
        ): LazyFrame =
            let cols = columns |> Seq.cast<IColumnExpr>
            this.Sort(cols, descending, nullsLast, ?maintainOrder = maintainOrder)
        member this.Sort(
            columns: seq<IColumnExpr>,
            ?descending: bool,
            ?nullsLast: bool,
            ?maintainOrder: bool
        ) : LazyFrame =
            let desc = defaultArg descending false
            let nLast = defaultArg nullsLast false
            this.Sort(columns, [| desc |], [| nLast |], ?maintainOrder = maintainOrder)
        member this.Sort(
            columns: seq<Expr>,
            ?descending: bool,
            ?nullsLast: bool,
            ?maintainOrder: bool
        ) : LazyFrame =
            let cols = columns |> Seq.cast<IColumnExpr>
            this.Sort(cols, ?descending = descending, ?nullsLast = nullsLast, ?maintainOrder = maintainOrder)
        member this.Sort(expr: Expr, ?descending: bool, ?nullsLast: bool) : LazyFrame =
            this.Sort([expr :> IColumnExpr], ?descending=descending, ?nullsLast=nullsLast)

        member this.Sort(colName: string, ?descending: bool, ?nullsLast: bool) : LazyFrame =
            this.Sort([Expr.Col colName :> IColumnExpr], ?descending=descending, ?nullsLast=nullsLast)

        member this.Orderby (expr: Expr, desc: bool) =
            this.Sort(expr, descending = desc)

    type DataFrame with
        /// <summary>
        /// Sort the DataFrame.
        /// </summary>
        /// <param name="columns">the column which needs to be sorted (Expr/Selector)。</param>
        /// <param name="descending">sort direction (true=descending).Length must be 1 (broadcasting) or same with columns.</param>
        /// <param name="nullsLast">null value position (true=last).Length must be 1 (broadcasting) or same with columns</param>
        /// <param name="maintainOrder">Stable Sort option</param>
        member this.Sort (
            columns: seq<IColumnExpr>,
            descending: seq<bool>,
            nullsLast: seq<bool>,
            ?maintainOrder: bool
        ): DataFrame =
            let lf = this.Lazy()
            lf.Sort(columns, descending, nullsLast, ?maintainOrder = maintainOrder).Collect()

        member this.Sort (
            columns: seq<Expr>,
            descending: seq<bool>,
            nullsLast: seq<bool>,
            ?maintainOrder: bool
        ): DataFrame =
            let cols = columns |> Seq.cast<IColumnExpr>
            this.Sort(cols, descending, nullsLast, ?maintainOrder = maintainOrder)

        /// <summary> Sort with simple broadcasting options. </summary>
        member this.Sort(
            columns: seq<IColumnExpr>,
            ?descending: bool,
            ?nullsLast: bool,
            ?maintainOrder: bool
        ) : DataFrame =
            let desc = defaultArg descending false
            let nLast = defaultArg nullsLast false
            this.Sort(columns, [| desc |], [| nLast |], ?maintainOrder = maintainOrder)
        member this.Sort(
            columns: seq<Expr>,
            ?descending: bool,
            ?nullsLast: bool,
            ?maintainOrder: bool
        ) : DataFrame =
            let cols = columns |> Seq.cast<IColumnExpr>
            this.Sort(cols, ?descending = descending, ?nullsLast = nullsLast, ?maintainOrder = maintainOrder)

        /// <summary> Sort by a single expression. </summary>
        member this.Sort(expr: Expr, ?descending: bool, ?nullsLast: bool) : DataFrame =
            this.Sort([expr :> IColumnExpr], ?descending=descending, ?nullsLast=nullsLast)

        /// <summary> Sort by a single column name. </summary>
        member this.Sort(colName: string, ?descending: bool, ?nullsLast: bool,?maintainOrder) : DataFrame =
            this.Sort([Expr.Col colName :> IColumnExpr], ?descending=descending, ?nullsLast=nullsLast,?maintainOrder = maintainOrder)
